### PR TITLE
Use mont reduced operations

### PR DESCRIPTION
**Pull request by kilic**
In kilic/bls12-381 lib `Fr` `Mul` operation is not the fastest way to compute multiplication. Instead we can use Montgomerry reduced operations.

Added under the `kilic_bignum_red` tag.

```

# go test -run none -bench BenchmarkFFTSettings_FFT -v

BenchmarkFFTSettings_FFT/scale_4-8                108186             10992 ns/op
BenchmarkFFTSettings_FFT/scale_5-8                 46240             24960 ns/op
BenchmarkFFTSettings_FFT/scale_6-8                 20463             55935 ns/op

# go test -run none -bench BenchmarkFFTSettings_FFT -v -tags bignum_kilic

BenchmarkFFTSettings_FFT/scale_4-8                233720              4910 ns/op
BenchmarkFFTSettings_FFT/scale_5-8                111338             10745 ns/op
BenchmarkFFTSettings_FFT/scale_6-8                 51358             23476 ns/op

# go test -run none -bench BenchmarkFFTSettings_FFT -v -tags bignum_kilic_red

BenchmarkFFTSettings_FFT/scale_4-8                432556              2792 ns/op
BenchmarkFFTSettings_FFT/scale_5-8                191559              6056 ns/op
BenchmarkFFTSettings_FFT/scale_6-8                 90892             13125 ns/op
```

### CODE REVIEW

File: bignum_kilic_red.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+func init() {
+	initGlobals()
+	ClearG1(&ZERO_G1)
+	initG1G2()
+}
+
+type Big kbls.Fr
+
+func norm(in *Big) *Big {
+	e := new(Big)
```
Comment (protolambda) [inline bignum_kilic_red.go:22]: Why create a new big-number? Can we not just run `.FromRed()` on the input, and define it as `func norm(v *Big)`, changing it in-place? If the user of the function wants a copy, they can always copy it themselves.

File: fft_g1.go [hunk 2]
```diff
@@ -12,13 +12,13 @@ func (fs *FFTSettings) simpleFTG1(vals []G1, valsOffset uint64, valsStride uint6
 	for i := uint64(0); i < l; i++ {
 		jv := &vals[valsOffset]
 		r := &rootsOfUnity[0]
-		mulG1(&v, jv, r)
+		mulG1(&v, jv, norm(r))
```
Comment (protolambda) [inline fft_g1.go:15]: Hmm, repeating `norm` for every root of unity seems like something that could be precomputed.

File: go.mod [hunk 3]
```diff
@@ -5,5 +5,6 @@ go 1.15
 require (
 	github.com/herumi/bls-eth-go-binary v0.0.0-20201019012252-4b463a10c225
 	github.com/holiman/uint256 v1.1.1
-	github.com/kilic/bls12-381 v0.0.0-20201226121925-69dacb279461
+	github.com/kilic/bls12-381 v0.0.0-20210111123357-c0f46a7d5c1f
```
Comment (protolambda) [inline go.mod:8]: Thanks, forgot to keep the deps updated. Might be better for another PR though. Do you plan to tag releases of your library? Or is there a changelog somewhere?

File: kzg_multi_proofs_test.go [hunk 4]
```diff
@@ -12,16 +11,16 @@ func TestKateSettings_CheckProofMulti(t *testing.T) {
 	s1, s2 := generateSetup("1927409816240961209460912649124", 16+1)
 	ks := NewKateSettings(fs, s1, s2)
 	for i := 0; i < len(ks.secretG1); i++ {
-		t.Logf("secret g1 %d: %s", i, strG1(&ks.secretG1[i]))
+		// t.Logf("secret g1 %d: %s", i, strG1(&ks.secretG1[i]))
```
Comment (protolambda) [inline kzg_multi_proofs_test.go:14]: The logs are here so I can compare double-check the values produced by the python implementation. See https://github.com/ethereum/research/tree/master/kzg_data_availability
Just keep them there, benchmarks are separate.

File: kzg_single_proofs_test.go [hunk 5]
```diff
@@ -11,6 +13,7 @@ func TestKateSettings_CheckProofSingle(t *testing.T) {
 	for i := 0; i < len(ks.secretG1); i++ {
 		t.Logf("secret g1 %d: %s", i, strG1(&ks.secretG1[i]))
 	}
+	t.Logf("secret g1 %d: %s", 3, strG1(&ks.secretG1[3]))
```
Comment (protolambda) [inline kzg_single_proofs_test.go:16]: Interesting, remove logs, then add one here? What's this one for?

File: bignum_kilic_red.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
+// +build bignum_kilic_red
```
Comment (protolambda) [inline bignum_kilic_red.go:1]: Thanks for the demo, testing different bignum implementations against each other is doable with the build-options. But this is a little bit too much copy-paste of the other kilic BLS file. We could either put the two `norm` function variants in separate files (and only switch that on/off with a build tag), or we just improve the existing kilic bignum file.

**Review by protolambda — COMMENTED** [review]
Hey! Thanks for your interest, your BLS library has been great. It may even become the default with all the improvements and non-CGO assembly usage. And where did you learn about this experimental Kate implementation?

Here are some review comments. I'll hold off with merging for now, I think we can polish the usage of `norm` after I understand it better, and the PR adds a lot of code, while the real change could probably be more isolated. Thanks a lot though, the performance improvement looks very promising!

**Comment by kilic** [comment]
> Hey! Thanks for your interest, your BLS library has been great. It may even become the default with all the improvements and non-CGO assembly usage. And where did you learn about this experimental Kate implementation?

Thanks :) I've been searching usage library of the library here github.

> while the real change could probably be more isolated

Right. Kept old `kilic_bignum` allow you to compare results easily.

Also planning to look at G1 add usage tomorrow. If second operand `b` in `(ret, b, a)` is affine point then it uses mixed addition which is slightly cheaper.

File: go.mod (hunk 3 above)
Comment (kilic) [reply go.mod:8]: This PR depends on very recent changes of the base library. [see](https://github.com/kilic/bls12-381/commit/c0f46a7d5c1fe4e1b185e6e6a0e3fe2041669607)

>  Do you plan to tag releases of your library?

Yes in two weeks just after a couple of improvements.

**Comment by protolambda** [comment]
Finally integrated this via #12, thank you :)
